runtime/near-vm-logic/src/logic.rs

```diff
@@ -2786,6 +2786,44 @@ impl<'a> VMLogic<'a> {
         }
         Ok(())
     }
+
+    /// Verify an ED25519 signature given a message and a public key.
+    /// # Cost
+    ///
+    /// `base + write_register_base + write_register_byte * num_bytes + ed25519_verify_base + ed25519_verify_byte * num_bytes`
```

Comment:
1. Could you explain the return value in the comment please?
2. I am not sure where these costs come from exactly. It's definitely not what is actually charged. Please update it here (and in the NEP) such that it reflects what is actually charged.
    *  Depending on whether a register or a pointer is provided, `get_vec_from_memory_or_register` charges different parameters.
    *  Input is read, so the costs should be `read_*` not `write_*`.
    * `num_bytes` is ambigous.

Reply:
1. done
2. I've changed it to 
```
/// `ed25519_verify_base + read_register_byte * (num_bytes_signature + num_bytes_message)`
```

Reply:
`read_register_base` should also be added.

And in the case of pointers instead of registers, it should be `read_memory_base` and `read_memory_bytes` respectively. It would be great to at least mention that in the comment,

Reply:
```
    /// `read_register_base + read_memory_base + read_memory_bytes * (num_bytes_signature + num_bytes_message) +
    ///  ed25519_verify_base + read_register_byte * (num_bytes_signature + num_bytes_message)`
```

Reply:
These are the correct parameters. But it's either one or the other that is charged for each memory or register read. Not both for everything.

There is a bit of magic going on inside the method `get_vec_from_memory_or_register` that you used in your implementation. Based on the length field, it could be a register or a memory location.
```
    fn get_vec_from_memory_or_register(&mut self, offset: u64, len: u64) -> Result<Vec<u8>> {
        if len != u64::MAX {
            self.memory_get_vec(offset, len)
        } else {
            self.internal_read_register(offset)
        }
    }
```
If you look at the source code, you will see that `memory_get_vec` charges gas based on memory parameters and `internal_read_register` based on registers.

Since the implementation uses `get_vec_from_memory_or_register`, the actual gas cost can be also be a combination. For example, memory costs for reading the signature, but register costs for reading the message and the public key.

Reply:
How would you like to express this on the docs? 
My attempt:
```rs
/// read_memory can be defined either as `read_memory_base + read_memory_bytes` if chunk of memory is bigger
/// than u64::MAX, otherwise it's `read_register_base + read_register_byte`
/// `read_memory * (num_bytes_signature + num_bytes_message, num_bytes_public_key) +
///  ed25519_verify_base + ed25519_verify_byte * (num_bytes_signature + num_bytes_message)`
```

Reply:
This doesn't quite sound right, it has nothing to do with the size of the memory. It's just that a special number as the length parameter signals that the input is a register, not a pointer.

Maybe let's integrate something like this in the comment:

_Each input can either be in memory or in a register. Set the length of the input to `u64::MAX` to declare that the input is a register number and not a pointer._

_Each input has a gas cost `input_cost(num_bytes)` that depends on whether it is from memory or from a register. It is either `read_memory_base + num_bytes * read_memory_byte` in the former case or `read_register_base + num_bytes * read_register_byte` in the latter._

Then you can use `input_cost(num_bytes)` in the actual formula.